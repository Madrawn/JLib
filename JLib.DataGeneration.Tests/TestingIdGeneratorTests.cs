using FluentAssertions;
using JLib.Helper;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;
using Xunit;

namespace JLib.DataGeneration.Tests;

public class TestingIdGeneratorTests : IDisposable
{
    class Nested<T>
    {
        private readonly TestingIdGenerator _idGenerator;

        public Nested(TestingIdGenerator idGenerator)
        {
            _idGenerator = idGenerator;
        }

        public Guid CreateId<T2, T3>()
            => CreateId<T2>(1);
        public Guid CreateId<T2>(int stackTraceFrameIndex)
            => _idGenerator.CreateGuid(stackTraceFrameIndex);
        public Guid CreateId<T2>(string stringParam)
            => CreateId<T2>(1);
    }

    private readonly List<IDisposable> _disposables = new();
    private readonly TestingIdGenerator _idGenerator;
    private readonly IServiceProvider _provider;


    public TestingIdGeneratorTests()
    {
        var provider = new ServiceCollection()
            .AddAutoMapper(cfg => { })
            .AddTestingIdGenerator()
            .AddIdRegistry(new() { DefaultNamespace = "JLib.DataGeneration.Tests" })
            .AddSingleton(typeof(Nested<>))
            .BuildServiceProvider();
        _disposables.Add(provider);
        _idGenerator = provider.GetRequiredService<TestingIdGenerator>();
        _provider = provider;

    }
    [Fact]
    public void CreateGuid_ShouldReturnNewGuid()
    {
        _idGenerator.CreateGuid().MatchSnapshot();
    }
    [Fact]
    public void CreateGuid_ShouldReturnNewGuid2()
    {
        _provider.GetRequiredService<Nested<int>>()
            .CreateId<double>(0)
            .MatchSnapshot();
    }
    [Fact]
    public void CreateGuid_ShouldReturnNewGuid3()
    {
        var a = _provider.GetRequiredService<Nested<int>>()
            .CreateId<double>(0);
        var b = _provider.GetRequiredService<Nested<int>>()
                .CreateId<double>(0);
        new[] { a, b }.MatchSnapshot();
        a.Should().NotBe(b);
    }
    [Fact]
    public void CreateGuid_ShouldReturnNewGuid4()
    {
        var a = _provider.GetRequiredService<Nested<int>>()
            .CreateId<double>(0);
        var b = _provider.GetRequiredService<Nested<int>>()
            .CreateId<double>(0);

        _idGenerator.SetIdScope(new("otherScope"));

        var c = _provider.GetRequiredService<Nested<int>>()
            .CreateId<double>(0);

        new[] { a, b, c }.MatchSnapshot();
        a.Should().NotBe(b).And.NotBe(c);
    }
    [Fact]
    public void CreateGuid_ShouldReturnNewGuid5()
    {
        var a = _provider.GetRequiredService<Nested<int>>()
            .CreateId<double>(0);
        var b = _provider.GetRequiredService<Nested<int>>()
            .CreateId<double, int>();
        new[] { a, b }.MatchSnapshot();
        a.Should().NotBe(b);
    }

    public void Dispose()
        => _disposables.DisposeAll();
}
