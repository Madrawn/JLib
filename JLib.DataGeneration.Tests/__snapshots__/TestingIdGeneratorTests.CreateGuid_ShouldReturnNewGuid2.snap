"74519d31-bb45-48c3-a7fa-f2e55a2c365f"
